### DISCUSSION TITLE
How to interpret the data (how to get started in general)

### DISCUSSION THREAD

**Discussion by Ben93kie** [semuconsulting/pyubx2#114] [Q&A] [answered]
Hi, 
first of all, thanks for the repo! I think this is really useful.

However, as I'm new to the GNSS world, I'm struggling a bit to start.

I got a F9P and F9H, which I want to leverage to get precise headings (on a moving boat, both devices on-boat).

I tried to play around with your repository, but could not get any useful output so far. I tried U-Center and there, I can at least read the correct lat and long values (but also do not know where/how to get the heading).

I tried

>>> stream = Serial('COM3',38400, timeout=3)
>>> ubr=UBXReader(stream)
>>> ubr.read()
(b'$GNRMC,110125.00,A,4831.47433,N,00902.54081,E,0.036,,100523,,,D,V*1D\r\n', NMEAMessage('GN','RMC', 0, payload=['110125.00', 'A', '4831.47433', 'N', '00902.54081', 'E', '0.036', '', '100523', '', '', 'D', 'V']))
>>> ubr.read()
(b'$GNVTG,,T,,M,0.036,N,0.067,K,D*3C\r\n', NMEAMessage('GN','VTG', 0, payload=['', 'T', '', 'M', '0.036', 'N', '0.067', 'K', 'D']))
>>> ubr.read()
(b'$GNGGA,110125.00,4831.47433,N,00902.54081,E,2,12,0.63,343.2,M,47.5,M,,0000*4B\r\n', NMEAMessage('GN','GGA', 0, payload=['110125.00', '4831.47433', 'N', '00902.54081', 'E', '2', '12', '0.63', '343.2', 'M', '47.5', 'M', '', '0000']))

But I do not know how to interpret this. Where is the lat, where is the long? How would I safe these values to variables? Finally, how would I get the heading, which is supposed to be in the UBX-NAV-RELPOSNED message (what does that even mean, how would I access this message)?

I also tried your webserver, but there the lat, long values are completely off and change completely also.

Sorry for the amount of questions, I'd be happy if you could answer some of them.

-Benjamin

**Comment by semuadmin** [comment]
Hi @Ben93kie 

pyubx2 is a Python library which is intended to be used within a user's own Python script or application, so it assumes some knowledge of Python and the NMEA, UBX & RTCM protocols.

The `UBXReader.read()` method returns a tuple of (raw data, parsed data). The parsed data is either a UBXMessage, NMEAMessage or RTCMMessage object, depending on the incoming message protocol - in your case it'll be an NMEAMessage. To view the parsed data in a human-readable format, simply use the `print(parsed)` statement (or the `str(parsed)` function). You can also access individual message attributes like lat, lon or alt, e.g.

```
>>> raw, parsed = ubr.read()
>>> print(parsed)
<NMEA(GNGGA, time=11:01:25, lat=48.5245721667, NS=N, lon=9.0423468333, EW=E, quality=2, numSV=12, HDOP=0.63, alt=343.2, altUnit=M, sep=47.5, sepUnit=M, diffAge=, diffStation=0)>
>>> print(f"lat {parsed.lat}, lon {parsed.lon}, alt {parsed.alt}")
lat 48.5245721667, lon 9.0423468333, alt 343.2
```

There are plenty of code examples in the [examples](https://github.com/semuconsulting/pyubx2/tree/master/examples) folder. There's also comprehensive [API documentation](https://www.semuconsulting.com/pyubx2/) available.

You mentioned you tried to use the 'webserver' - I presume you're referring to [this example](https://github.com/semuconsulting/pyubx2/tree/master/examples/webserver)? Note that this example only works with UBX data - specifically UBX NAV-DOP, NAV-POSLLH, NAV-HPPOSLLH, NAV-PVT and NAV-SAT message types. By default, a u-blox receiver does not output UBX protocol data - it only outputs NMEA data, which is what you're seeing above. For the webserver example to work, you will need to enable UBX output. Refer to the u-blox documentation for how to do this, or refer to some of the other examples provided.

**Comment by Ben93kie** [comment]
Thanks for the quick reply!
Yes, I'm reading into the NMEA (and also UBX) protocols currently. 

I got the values to print out the correct lat and long via this simple script (essentially your example in the README):

from serial import Serial
from pyubx2 import UBXReader
stream = Serial('COM3', 38400, timeout=1)
ubr = UBXReader(stream)
(raw_data, parsed_data) = ubr.read()
print(parsed_data)

The message I receive is this

<NMEA(GNRMC, time=23:00:09, status=A, lat=48.5344915, NS=N, lon=9.0706453333, EW=E, spd=0.147, cog=, date=2023-05-10, mv=, mvEW=, posMode=D, navStatus=V)>

I correctly get the lat and long, but I do not see any "heading" output. Could you point me to somewhere where I would be able to implement that eventually? Would I have to use your other repo (https://github.com/semuconsulting/pynmeagps) or is this already fully included here? 
I'm afraid that I have to look into the low-level stuff of  RTCM corrections to achieve that, but I'm trying to avoid that.

Yes, the webserver explanations make sense to me, too. Thanks!

**Comment by semuadmin** [comment, marked as answer]
Hi @Ben93kie 

Using the F9P and F9H receivers in the recommended moving base/rover configuration is a non-trivial exercise requiring fairly detailed knowledge of the protocols involved.

_**Please appreciate that I do not work for, and am not paid by, u-blox. These libraries are maintained entirely voluntarily, though donations are always appreciated.**_

I recommend you do some basic research into GNSS NMEA and UBX message protocols - there are plenty of resources you can find using Google. The u-blox Interface Manuals for your F9P and F9H receivers contain a wealth of information on the messages they support:

https://www.u-blox.com/sites/default/files/documents/u-blox-F9-HPG-1.32_InterfaceDescription_UBX-22008968.pdf
https://content.u-blox.com/sites/default/files/u-blox_ZED-F9H_InterfaceDescription_%28UBX-19030118%29.pdf

You can ask questions on the u-blox Support Forum here: https://portal.u-blox.com/s/

I also publish this one-page [GNSS Positioning - A Reviser](https://www.semuconsulting.com/gnsswiki/) which gives an overview of the technology.

In short, a GNSS receiver in its default configuration outputs various NMEA message types, each of which contains a specific set of data items. An NMEA message identifier comprises a one or two character 'talker' (which corresponds to the GNSS constellation) and a three character 'message ID' (which denotes the message type and content) e.g. GNRMC, where 'GN' signifies GNSS and 'RMC' signifies 'Recommended Minimum Content'.

By default, a u-blox F9P receiver will output the following types at a rate of approximately one a second:

Talkers:
    "GA": "Galileo Positioning System",
    "GB": "BDS (BeiDou System) ",
    "GI": "NavIC (IRNSS)",
    "GL": "GLONASS Receiver",
    "GN": "Global Navigation Satellite System (GNSS)", (i.e. generic)
    "GP": "Global Positioning System (GPS)",
    "GQ": "QZSS",
    "P": "Proprietary",

Message IDs:
    "GGA": "Global positioning system fix data",
    "GLL": "Latitude and longitude, with time of position fix and status",
    "GSA": "GNSS Dilution of Precision (DOP) and Active Satellites",
    "GSV": "GNSS Satellites in View",
    "RMC": "Recommended Minimum Content", (a kind of one-stop summary)
    "VTG": "Course over ground and Groundspeed",

So if, for example, you want course over ground ('cog', or 'cogt'/'cogm' (true/magnetic))  you can find that in the RMC or VTG messages. Bear in mind a basic, unassisted GNSS receiver cannot give you 'heading' as such, only track or course over ground. The F9P and F9H receivers are designed to work in tandem to give you heading, using a technique known as RTK (real-time kinematics).

To find out which type a given message is, using the 'identity' attribute:

ubr = UBXReader(stream)
(raw_data, parsed_data) = ubr.read()
print(parsed_data.identity)
if parsed_data.identity == "GNRMC": print(parsed_data.cog)

**NB:** NMEA is a legacy protocol which was designed before modern GNSS capabilities were introduced. u-blox receivers also support the proprietary UBX protocol which covers a far more comprehensive range of message types and data items. For sophisticated applications, you may need to enable the UBX protocol as well as, or instead of, NMEA. Refer to the receiver's interface and integration manuals for more information.

**Reply by Ben93kie** [reply]
Thanks for your comprehensive reply!
I didn't think that it would be that great of a rabbit hole to just get the heading information for me as an end user. Maybe, I should have bought a different module/dev kit that already provides that via some easy examples :D

[I asked in the forum](https://portal.u-blox.com/s/question/0D72p00000CbQoaCAF/detail?fromEmail=1&s1oid=00D200000001ogf&s1nid=0DB2p0000008PMD&s1uid=0052p00000CDCvO&s1ext=0&emkind=chatterCommentNotification&emtm=1683772355592&t=1683807804473), but this still does not even work in U-Center. 

Maybe, I will make it eventually.

Thank you in any way, your help is greatly appreciated!

**Reply by semuadmin** [reply]
Just to be clear, getting the track (course over ground) data from NMEA output is trivial. Almost any basic GNSS receiver module will give you that.

But you've presumably acquired your F9P and F9H modules (_at not inconsiderable expense!_) with the intention of using them in a moving base/rover RTK configuration? That's an ambitious thing to be doing if you're "new to the GNSS world", but the information you need is all in the u-blox integration and interface manuals - you just need to do the reearch!

You might want to get your head around some basic GNSS concepts e.g. using pyubx2 with the F9P as a stand-alone receiver, before embarking on an multi-receiver project.

**Reply by Ben93kie** [reply]
Yes, moving base-rover config. And yes, expensive and complicated :D

You're right, I have to look at the basics first. I just have to admit that it takes time.
Dropped a small donation for the helpful response, thanks!
